lib/Dialect/AIEVec/Transforms/ConvertVectorToAIEVec.cpp

```diff
@@ -562,6 +568,586 @@ struct SetInboundsToReadStoreOpPattern : public RewritePattern {
+        } else {
+          if (!indexToExprDimMap.count(index))
+            indexToExprDimMap[index] =
+                getAffineDimExpr(indexToExprDimMap.size(), context);
+          indexExprs.push_back(indexToExprDimMap[index]);
+        }
+      }
+
+      exprVec.push_back(map.getResult(0).replaceDims(indexExprs));
+    } else if (auto cOp = value.getDefiningOp<arith::ConstantOp>()) {
+      auto idxVal = cOp.getValue().cast<IntegerAttr>().getValue();
+      unsigned idx = idxVal.getSExtValue();
+      exprVec.push_back(getAffineConstantExpr(idx, context));
+    } else {
+      if (!indexToExprDimMap.count(value))
+        indexToExprDimMap[value] =
+            getAffineDimExpr(indexToExprDimMap.size(), context);
+      exprVec.push_back(indexToExprDimMap[value]);
+    }
+  }
+
+  assert(!exprVec.empty() && "Could not construct linearized affineExpr");
+
+  auto ret = makeFlattenedStridedExpr(memRefType.getShape(), exprVec,
+                                      memRefType.getContext());
+
+  return ret;
+}
+
+// From a linearized affine expression, compute the base and the constant
+// offset. If the access is A[i][j+2] for an N*N array A, the linearized
+// expression will be A[i*N+j+2]. The base in this case will be (i*N+j), and the
+// offset will be 2.
+static std::pair<AffineExpr, int32_t> getBaseAndOffset(AffineExpr expr) {
+  AffineExpr base = expr;
+  int32_t offset = 0;
+
+  if (auto constExpr = expr.dyn_cast<AffineConstantExpr>()) {
+    base = nullptr;
+    offset += constExpr.getValue();
+  } else if (auto binopExpr = expr.dyn_cast<AffineBinaryOpExpr>()) {
+    if (binopExpr.getKind() == AffineExprKind::Add) {
+      AffineExpr lhs = binopExpr.getLHS(), rhs = binopExpr.getRHS();
+      if (auto constExpr = lhs.dyn_cast<AffineConstantExpr>()) {
+        base = rhs;
+        offset += constExpr.getValue();
+      }
+      if (auto constExpr = rhs.dyn_cast<AffineConstantExpr>()) {
+        base = base == rhs ? nullptr : lhs;
+        offset += constExpr.getValue();
+      }
+    }
+  }
+  return std::make_pair(base, offset);
+}
+
+// If the defining op of one of add operands is an add op, return this defining
+// add op.
+static arith::AddIOp getDefAddOp(arith::AddIOp addOp) {
+  arith::AddIOp defLhs =
+      dyn_cast<arith::AddIOp>(addOp->getOperand(0).getDefiningOp());
+  arith::AddIOp defRhs =
+      dyn_cast<arith::AddIOp>(addOp->getOperand(1).getDefiningOp());
+  if ((!defLhs && !defRhs) || (defLhs && defRhs)) {
+    return nullptr;
+  }
+  return defLhs ? defLhs : defRhs;
+}
+
+// The mul add chain pattern is valid if the defining ops of a mul op consist of
+// a broadcast op and an upd op.
+static bool checkChainPattern(arith::MulIOp mulOp, MulDefMapTy &macChainMap,
+                              SmallVectorImpl<Value> &bcastOpSourceVec) {
+  // Get the mul op's lhs and rhs defining ops. We keep splat op at
+  // rhs.
+  if (isa<aievec::BroadcastOp>(mulOp.getOperand(0).getDefiningOp())) {
```

Comment:
checkChainPattern must not modify the IR for cases where the pattern does not apply!